76 barzahlungcontroller implementieren

Benokit4 opened the pull request:
BarzahlungController implementiert, sodass die geforderte Funktionalität gegeben ist.

Benokit4 committed:
BarzahlungController implementiert

Funktioniert gemäß den Anforderungen

Benokit4 committed:
BarzahlungView angepasst

Getter-Methoden hinzugefügt

Benokit4 committed:
Geldbetrag greaterThanOrEqualTo hinzugefügt

Benokit4 committed:
Update PlatzVerkaufsController.java

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungController.java:
```diff
@@ -0,0 +1,137 @@
+	private void updateRestbetrag() 
+	{
+		String einzahlungText = _view.getEinzahlungTextField().getText();
+        Geldbetrag einzahlung = Geldbetrag.fromString(einzahlungText).orElse(Geldbetrag.ZERO);
```
Sollte nicht Geldbetrag.ZERO sein, falls das nicht klappt, sondern der letzte gültige Geldbetrag! (Sofern bisher kein gültiger Geldbetrag eingegeben wurde, ist dies Geldbetrag.ZERO)

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungController.java:
```diff
@@ -0,0 +1,137 @@
+		String einzahlungText = _view.getEinzahlungTextField().getText();
+        Geldbetrag einzahlung = Geldbetrag.fromString(einzahlungText).orElse(Geldbetrag.ZERO);
+        Geldbetrag restbetrag = _preis.subtract(einzahlung);
+        _view.getRestbetragLabel().setText(restbetrag.toString() + "€");
```
Was passiert bei negativem Restbetrag?

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungView.java:
```diff
@@ -0,0 +1,207 @@
+                SwingConstants.CENTER);
+        JLabel textRestbetragLabel = new JLabel("Restbetrag",
+                SwingConstants.CENTER);
+        _restbetragLabel = new JLabel("00,00€", SwingConstants.CENTER);
```
Magischer Wert "00,00€", falls Geldbetrag.toString() verändert wird, müsste auch dieser magische Wert verändert werden! Passt nicht zu Prinzip, dass Änderungen lokal bleiben sollte, schlechtere Wartbarkeit.
Z.b. falls von Euro auf USD gewechselt wird.

frndotmsn replied:
Eher im Controller machen, und dann Geldbetrag.ZERO.toString() oder so mit Währung als String formatieren und an das zugehörige JLabel weitergeben

Benokit4 committed:
Abbrechen Funktionalität überarbeitet

Nun lässt sich jederzeit der Abbrechen-Button betätigen, unabhängig von der Eingabe.

Benokit4 committed:
Lesbarkeit registriereUIAktionen() verbessert

Benokit4 committed:
Fehlermeldung eingefügt

Fehlermeldungen eventuell spezifischer gestalten

Benokit4 committed:
_fehlermeldungLabel hinzugefügt

Redundantes _preisLabel entfernt

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungView.java:
```diff
@@ -0,0 +1,207 @@
+        _dialog = new JDialog();
+        _dialog.setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
+        _dialog.setResizable(false);
+        _dialog.setModal(false);
```
eigentlich wollten wir modal, oder? damit nicht mehrere Barzahlungsdialoge auf einmal geöffnet werden können

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungController.java:
```diff
@@ -0,0 +1,137 @@
+	 * @param view Die Ansicht für die Barzahlung.
+	 * @param vorstellung Die Vorstellung, fuer die die Plaetze verkauft werden sollen.
+	 */
+	public BarzahlungController(JPlatzplan platzplan, BarzahlungView view, Vorstellung vorstellung)
```
Warum wird view extern erstellt und weitergegeben? PlatzVerkaufsController erstellt auch selbst die PlatzVerkaufsView

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/platzverkauf/PlatzVerkaufsController.java:
```diff
@@ -161,27 +164,90 @@ private void aktualisierePlatzplan()
+    private void zeigeBarzahlungsDialog(Set<Platz> plaetze, JPlatzplan platzplan) 
+    {
+    	Geldbetrag preis = _vorstellung.getPreisFuerPlaetze(plaetze);
+        BarzahlungView barzahlungView = new BarzahlungView(preis);
```
Nicht aufgabenbereich vom PlatzVerkaufsController die BarzahlungsView zu erstellen (WAM und MVC -> Jeder Controller ist nur für seine eigene View zuständig und die Koordinierung mit anderen Controllern)

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungView.java:
```diff
@@ -0,0 +1,207 @@
+     * Initialisert die Oberfläche. Die Parameter sind die Views der Submodule,
+     * die eingebettet werden.
+     */
+    public BarzahlungView(Geldbetrag geldbetrag)
```
Warum nimmt View Parameter geldbetrag, ist das nicht Aufgabenbereich des Controllers, dass der Geldbetrag direkt an das entsprechende JLabel weitergegeben wird?

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/kino/ui/barzahlung/BarzahlungView.java:
```diff
@@ -0,0 +1,207 @@
+        JPanel barzahlungPanel = new JPanel();
+        barzahlungPanel.setLayout(new GridLayout(2, 3));
+        JLabel textPreisLabel = new JLabel("Preis", SwingConstants.CENTER);
+        _preisWertLabel = new JLabel(geldbetrag.toString(),
```
Durch das mehrfache Durchreichen von Geldbetrag wird hier ein Währungszeichen, in unserem Falle € vergessen! Lieber direkt im Controller machen, denn um die Währung soll sich die View nicht kümmern

frndotmsn reviewed (changes requested):
Hier müssen sicherlich noch ein paar Änderungen vorgenommen werden, außer dem, was ich angemerkt habe, sieht es aber sehr gut aus (z.B. deskriptive Namen, gutes Javadoc)!

frndotmsn commented:
#84 Umsetzung der Vorschläge
